File: sample/java/03_julia_set.java
final class _03_julia_set {
    private _03_julia_set() {
    }


    // 03: Sample that outputs a Julia set as a PNG image.
    // Implemented with simple loop-centric logic for transpilation compatibility.

    public static java.util.ArrayList<Long> render_julia(long width, long height, long max_iter, double cx, double cy) {
        java.util.ArrayList<Long> pixels = new java.util.ArrayList<Long>();
        for (long y = 0L; y < height; y += 1L) {
            double zy0 = (-(1.2)) + 2.4 * (((double)(y)) / (((double)(height - 1L))));
            for (long x = 0L; x < width; x += 1L) {
                double zx = (-(1.8)) + 3.6 * (((double)(x)) / (((double)(width - 1L))));
                double zy = zy0;
                long i = 0L;
                while (((i) < (max_iter))) {
                    double zx2 = zx * zx;
                    double zy2 = zy * zy;
                    if (((zx2 + zy2) > (4.0))) {
                        break;
                    }
                    zy = 2.0 * zx * zy + cy;
                    zx = zx2 - zy2 + cx;
                    i += 1L;
                }
                long r = 0L;
                long g = 0L;
                long b = 0L;
                if (((i) >= (max_iter))) {
                    r = 0L;
                    g = 0L;
                    b = 0L;
                } else {
                    double t = ((double)(i)) / ((double)(max_iter));
                    r = PyRuntime.__pytra_int(255.0 * (0.2 + 0.8 * t));
                    g = PyRuntime.__pytra_int(255.0 * (0.1 + 0.9 * (t * t)));
                    b = PyRuntime.__pytra_int(255.0 * (1.0 - t));
                }
                pixels.add(r);
                pixels.add(g);
                pixels.add(b);
            }
        }
        return pixels;
    }

    public static void run_julia() {
        long width = 3840L;
        long height = 2160L;
        long max_iter = 20000L;
        String out_path = "sample/out/03_julia_set.png";
        double start = time.perf_counter();
        java.util.ArrayList<Long> pixels = render_julia(width, height, max_iter, (-(0.8)), 0.156);
        png.write_rgb_png(out_path, width, height, pixels);
        double elapsed = time.perf_counter() - start;
        System.out.println(String.valueOf("output:") + " " + String.valueOf(out_path));
        System.out.println(String.valueOf("size:") + " " + String.valueOf(width) + " " + String.valueOf("x") + " " + String.valueOf(height));
        System.out.println(String.valueOf("max_iter:") + " " + String.valueOf(max_iter));
        System.out.println(String.valueOf("elapsed_sec:") + " " + String.valueOf(elapsed));
    }
}

File: sample/julia/03_julia_set.jl
include(joinpath(@__DIR__, "built_in", "py_runtime.jl"))

include(joinpath(@__DIR__, "std", "time.jl"))
include(joinpath(@__DIR__, "utils", "png.jl"))
png = (_adler32=_adler32, _chunk=_chunk, _crc32=_crc32, _png_append_list=_png_append_list, _png_u16le=_png_u16le, _png_u32be=_png_u32be, _zlib_deflate_store=_zlib_deflate_store, write_rgb_png=write_rgb_png)

# 03: Sample that outputs a Julia set as a PNG image.
# Implemented with simple loop-centric logic for transpilation compatibility.

function render_julia(width, height, max_iter, cx, cy)
    pixels = UInt8[]
    
    for y in 0:height - 1
        zy0 = ((-1.2) + (2.4 * (y / (height - 1))))
        
        for x in 0:width - 1
            zx = ((-1.8) + (3.6 * (x / (width - 1))))
            zy = zy0
            
            i = 0
            while (i < max_iter)
                zx2 = (zx * zx)
                zy2 = (zy * zy)
                if ((zx2 + zy2) > 4.0)
                    break
                end
                zy = (((2.0 * zx) * zy) + cy)
                zx = ((zx2 - zy2) + cx)
                i = i + 1
            end
            r = 0
            g = 0
            b = 0
            if (i >= max_iter)
                r = 0
                g = 0
                b = 0
            else
                t = (i / max_iter)
                r = __pytra_int((255.0 * (0.2 + (0.8 * t))))
                g = __pytra_int((255.0 * (0.1 + (0.9 * (t * t)))))
                b = __pytra_int((255.0 * (1.0 - t)))
            end
            push!(pixels, r)
            push!(pixels, g)
            push!(pixels, b)
        end
    end
    return pixels
end

function run_julia()
    width = 3840
    height = 2160
    max_iter = 20000
    out_path = "sample/out/03_julia_set.png"
    
    start = perf_counter()
    pixels = render_julia(width, height, max_iter, (-0.8), 0.156)
    png.write_rgb_png(out_path, width, height, pixels)
    elapsed = (perf_counter() - start)
    
    __pytra_print("output:", out_path)
    __pytra_print("size:", width, "x", height)
    __pytra_print("max_iter:", max_iter)
    __pytra_print("elapsed_sec:", elapsed)
end


run_julia();
